sound/soc/sof/intel/hda-codec.c

```diff
@@ -173,24 +174,23 @@ static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
 		 * other return codes without modification
 		 */
 		if (ret == 0)
-			goto error;
+			ret = -ENOENT;
 	}
 
-	return ret;
-
-error:
-	snd_hdac_ext_bus_device_exit(hdev);
-	return -ENOENT;
-
+out:
+	if (ret < 0) {
+		snd_hdac_device_unregister(hdev);
+		put_device(&hdev->dev);
```

Comment:
I am not sure if we need `put_device(&hdev->dev);`. 
`device_del` will be called by `snd_hdac_device_unregister` and I suppose `device_del` will handle device reference counting as `device_add` does.

I added some trace below
```patch
--- a/drivers/base/core.c
+++ b/drivers/base/core.c
@@ -3020,6 +3020,9 @@ EXPORT_SYMBOL_GPL(device_register);
  */
 struct device *get_device(struct device *dev)
 {
+       if (dev) {
+               dev_info(dev, "bard: %s dev:%p\n", __func__, dev);
+       }
        return dev ? kobj_to_dev(kobject_get(&dev->kobj)) : NULL;
 }
 EXPORT_SYMBOL_GPL(get_device);
@@ -3030,6 +3033,9 @@ EXPORT_SYMBOL_GPL(get_device);
  */
 void put_device(struct device *dev)
 {
+       if (dev) {
+               dev_info(dev, "bard: %s dev:%p\n", __func__, dev);
+       }
        /* might_sleep(); */
        if (dev)
                kobject_put(&dev->kobj);

```
and forced `ret = -ENOENT;` and remove the `put_device(&hdev->dev);` by
```patch
--- a/sound/soc/sof/intel/hda-codec.c
+++ b/sound/soc/sof/intel/hda-codec.c
@@ -178,9 +178,10 @@ static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
        }

 out:
+       ret = -ENOENT;
        if (ret < 0) {
                snd_hdac_device_unregister(hdev);
-               put_device(&hdev->dev);
+       //      put_device(&hdev->dev);
        }
 #else
        hdev = devm_kzalloc(sdev->dev, sizeof(*hdev), GFP_KERNEL);
```
I get the following trace log
```
[    7.808566] hdaudio ehdaudio0D2: bard: get_device dev:0000000062665a98
[    7.808573] sof-audio-pci 0000:00:1f.3: bard: get_device dev:00000000a04f5e63
[    7.808887] hdaudio ehdaudio0D2: bard: get_device dev:0000000062665a98
[    7.809057] hdaudio ehdaudio0D2: bard: get_device dev:0000000062665a98
[    7.809062] hdaudio ehdaudio0D2: bard: put_device dev:0000000062665a98
[    7.813079] snd_hda_codec_hdmi ehdaudio0D2: bard: put_device dev:0000000062665a98
[    7.813578] snd_hda_codec_hdmi ehdaudio0D2: bard: put_device dev:0000000062665a98
[    7.813744] sof-audio-pci 0000:00:1f.3: bard: put_device dev:00000000a04f5e63
```

Reply:
I think we still need to call put_device() because we call device_initialize() before.

Reply:
> I think we still need to call put_device() because we call device_initialize() before.

@libinyang My point is that `put_device()` is called by `snd_hdac_device_unregister() -> device_del() -> put_device()` already.
So, we may not need to call  `put_device()` after `snd_hdac_device_unregister()`.

Reply:
@bardliao Thanks for picking this up. I need to do more investigation. I have the put_device() there as I wanted to align with legacy flow. In hdac_ext_bus.c the extra put_device() is there:

```
void snd_hdac_ext_bus_device_remove(struct hdac_bus *bus)
{
»       struct hdac_device *codec, *__codec;
»       /*                                                                                                                                                                  
»        * we need to remove all the codec devices objects created in the                                                                                                   
»        * snd_hdac_ext_bus_device_init                                                                                                                                     
»        */
»       list_for_each_entry_safe(codec, __codec, &bus->codec_list, list) {
»       »       snd_hdac_device_unregister(codec);
»       »       put_device(&codec->dev);
»       }
}
```

... so I think there are two references taken. One when device is created and another when the device is registered to a device hiearchy. My understanding the put_device() in snd_hdac_device_unregister() releases only one of these, and thus the extra put_device() is needed. 

But hmm, I need to double-check your traces. I think this is explained as device_initialize() takes the first reference without calling get_device(). So I think in above case, one reference is still left and the device is not destroyed.

Reply:
@bardliao In device_initialize(), it will set the refcount to 1 without calling get_device(). So I think we should also call put_device() to make the count to 0, just like @kv2019i 's example of snd_hdac_ext_bus_device_remove(), it will call snd_hdac_device_unregister() and then put the device again.

Reply:
> @bardliao In device_initialize(), it will set the refcount to 1 without calling get_device(). So I think we should also call put_device() to make the count to 0, just like @kv2019i 's example of snd_hdac_ext_bus_device_remove(), it will call snd_hdac_device_unregister() and then put the device again.

Thanks @libinyang you convinced me. I checked the refcount is 1 after `snd_hdac_device_unregister(hdev);`. So the `put_device(&hdev->dev);` is needed.